src/inet/UDPEndPointImpl_OpenThread.cpp

```diff
@@ -0,0 +1,268 @@
+
+    ep->Retain();
+    CHIP_ERROR err = ep->GetSystemLayer().ScheduleLambda([ep, p = System::LwIPPacketBufferView::UnsafeGetLwIPpbuf(payload)] {
+        ep->HandleDataReceived(System::PacketBufferHandle::Adopt(p));
+        ep->Release();
+    });
+    if (err == CHIP_NO_ERROR)
+    {
+        // If ScheduleLambda() succeeded, it has ownership of the buffer, so we need to release it (without freeing it).
+        static_cast<void>(std::move(payload).UnsafeRelease());
+    }
+    else
+    {
+        ep->Release();
+    }
+}
+
+CHIP_ERROR UDPEndPointImplOT::IPv6Bind(otUdpSocket & socket, const IPAddress & address, uint16_t port, InterfaceId interface)
+{
+    (void) interface;
+    otError err = OT_ERROR_NONE;
+    otSockAddr listenSockAddr;
+
+    memset(&socket, 0, sizeof(socket));
+    memset(&listenSockAddr, 0, sizeof(listenSockAddr));
+
+    listenSockAddr.mPort    = port;
+    listenSockAddr.mAddress = chip::DeviceLayer::Internal::ToOpenThreadIP6Address(address);
+
+    otUdpOpen(mOTInstance, &socket, handleUdpReceive, this);
+    otUdpBind(mOTInstance, &socket, &listenSockAddr, OT_NETIF_THREAD);
+
+    return chip::DeviceLayer::Internal::MapOpenThreadError(err);
+}
+
+CHIP_ERROR UDPEndPointImplOT::BindImpl(IPAddressType addressType, const IPAddress & addr, uint16_t port, InterfaceId interface)
+{
+
+    if (addressType == IPAddressType::kIPv6)
+    {
+        ReturnErrorOnFailure(IPv6Bind(mSocket, addr, port, interface));
+    }
+    else
+    {
+        return INET_ERROR_WRONG_ADDRESS_TYPE;
+    }
+
+    mBoundPort   = port;
+    mBoundIntfId = interface;
+
+    return CHIP_NO_ERROR;
+}
+
+InterfaceId UDPEndPointImplOT::GetBoundInterface() const
+{
+    return mBoundIntfId;
+}
+
+uint16_t UDPEndPointImplOT::GetBoundPort() const
+{
+    return mBoundPort;
+}
+
+CHIP_ERROR UDPEndPointImplOT::ListenImpl()
+{
+    // Nothing to do. Callback was set upon Binding call.
+    return CHIP_NO_ERROR;
+}
+
+void UDPEndPointImplOT::HandleDataReceived(System::PacketBufferHandle && msg)
+{
+    if ((mState == State::kListening) && (OnMessageReceived != nullptr))
+    {
+        const IPPacketInfo * pktInfo = GetPacketInfo(msg);
+
+        if (pktInfo != nullptr)
+        {
+            const IPPacketInfo pktInfoCopy = *pktInfo; // copy the address info so that the app can free the
+                                                       // PacketBuffer without affecting access to address info.
+
+            msg->ConsumeHead(sizeof(IPPacketInfo));
+            OnMessageReceived(this, std::move(msg), &pktInfoCopy);
+        }
+        else
+        {
+            if (OnReceiveError != nullptr)
+            {
+                OnReceiveError(this, CHIP_ERROR_INBOUND_MESSAGE_TOO_BIG, nullptr);
+            }
+        }
+    }
+}
+
+CHIP_ERROR UDPEndPointImplOT::SetMulticastLoopback(IPVersion aIPVersion, bool aLoopback)
+{
+    (void) aIPVersion;
+    (void) aLoopback;
+    // TODO
+    return CHIP_NO_ERROR;
```

Comment:
should we return ERR_NOT_IMPLEMENTED here and below rather than silent do nothing?

If not, a comment explaining would help. `TODO` does not explain why NO_ERROR was  chosen.

Reply:
Set to ERR_NOT_IMPLEMENTED